contracts/extensions/ExpenditureUtils.sol

```diff
@@ -0,0 +1,222 @@
+
+  You should have received a copy of the GNU General Public License
+  along with The Colony Network. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+pragma solidity 0.7.3;
+pragma experimental ABIEncoderV2;
+
+import "./../colony/ColonyDataTypes.sol";
+import "./../colonyNetwork/IColonyNetwork.sol";
+import "./../patriciaTree/PatriciaTreeProofs.sol";
+import "./ColonyExtensionMeta.sol";
+
+// ignore-file-swc-108
+
+
+contract ExpenditureUtils is ColonyExtensionMeta, PatriciaTreeProofs {
+
+  // Storage
+
+  uint256 stakeFraction;
+
+  mapping (address => mapping (uint256 => uint256)) stakes;
+
+  // Modifiers
+
+  modifier onlyRoot() {
+    require(colony.hasUserRole(msgSender(), 1, ColonyDataTypes.ColonyRole.Root), "evaluated-expenditure-caller-not-root");
+    _;
+  }
+
+  // Overrides
+
+  /// @notice Returns the identifier of the extension
+  function identifier() public override pure returns (bytes32) {
+    return keccak256("ExpenditureUtils");
+  }
+
+  /// @notice Returns the version of the extension
+  function version() public override pure returns (uint256) {
+    return 1;
+  }
+
+  /// @notice Configures the extension
+  /// @param _colony The colony in which the extension holds permissions
+  function install(address _colony) public override auth {
+    require(address(colony) == address(0x0), "extension-already-installed");
+
+    colony = IColony(_colony);
+  }
+
+  /// @notice Called when upgrading the extension
+  function finishUpgrade() public override auth {}
+
+  /// @notice Called when deprecating (or undeprecating) the extension
+  function deprecate(bool _deprecated) public override auth {
+    deprecated = _deprecated;
+  }
+
+  /// @notice Called when uninstalling the extension
+  function uninstall() public override auth {
+    selfdestruct(address(uint160(address(colony))));
+  }
+
+  // Public
+
+  function setStakeFraction(uint256 _stakeFraction) public onlyRoot {
+    stakeFraction = _stakeFraction;
+  }
+
+  function makeExpenditureWithStake(
+    uint256 _permissionDomainId,
+    uint256 _childSkillIndex,
+    uint256 _domainId,
+    bytes memory _key,
+    bytes memory _value,
+    uint256 _branchMask,
+    bytes32[] memory _siblings
+  )
+    public
+  {
+    uint256 domainRep = getReputationFromProof(_domainId, _key, _value, _branchMask, _siblings);
+    uint256 stake = wmul(domainRep, stakeFraction);
+
+    colony.obligateStake(msgSender(), _domainId, stake);
+    colony.makeExpenditure(_permissionDomainId, _childSkillIndex, _domainId);
+
+    uint256 expenditureId = colony.getExpenditureCount();
+    stakes[msgSender()][expenditureId] = stake;
+    colony.transferExpenditure(expenditureId, msgSender());
+  }
+
+  function slashStake(
+    uint256 _permissionDomainId,
+    uint256 _childSkillIndex,
+    uint256 _expenditureId
+  )
+    public
+  {
+    ColonyDataTypes.Expenditure memory expenditure = colony.getExpenditure(_expenditureId);
+    uint256 stake = stakes[expenditure.owner][_expenditureId];
```

Comment:
Does this let all the necessary flows actually happen? It seems like a common ordering would be to give the ownership to someone else to manage the task, and then slash the original creator? Or indeed if ownership is transferred just with the natural flow of things, but the new owner misbehaves...

Reply:
It's a good question. Once the ownership is transferred, we can't really guard against it being transferred again. I'm comfortable saying that the original owner is on the hook if they transfer. That said, we should probably not check for ownership on the expenditure itself, but require it to be passed in (I had that originally, but then tried to be clever).